src/Sdk/Cerbos/Sdk/Cloud/V1/Interceptor/AuthInterceptor.cs

```diff
@@ -0,0 +1,152 @@
+        private ReaderWriterLock RWLock;
+
+        public AuthInterceptor(IApiKeyClient apiKeyClient, Credentials credentials)
+        {
+            ApiKeyClient = apiKeyClient;
+            Credentials = credentials;
+            RWLock = new ReaderWriterLock();
+        }
+
+        public override TResponse BlockingUnaryCall<TRequest, TResponse>(
+            TRequest request,
+            ClientInterceptorContext<TRequest, TResponse> context,
+            BlockingUnaryCallContinuation<TRequest, TResponse> continuation
+        )
+        {
+            ThrowIfUnauthenticated();
+
+            if (Expired())
+            {
+                IssueToken();
+            }
+
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method,
+                context.Host,
+                context.Options.WithHeaders(MetadataWithAuthToken(context.Options.Headers))
+            );
+            return continuation(request, newContext);
+        }
+        
+        public override AsyncUnaryCall<TResponse> AsyncUnaryCall<TRequest, TResponse>(
+            TRequest request,
+            ClientInterceptorContext<TRequest, TResponse> context,
+            AsyncUnaryCallContinuation<TRequest, TResponse> continuation)
+        {
+            ThrowIfUnauthenticated();
+
+            if (Expired())
+            {
+                IssueToken();
+            }
+
+            var newContext = new ClientInterceptorContext<TRequest, TResponse>(
+                context.Method,
+                context.Host,
+                context.Options.WithHeaders(MetadataWithAuthToken(context.Options.Headers))
+            );
+            return continuation(request, newContext);
+        }
+
+        private void IssueToken()
+        {
+            RWLock.AcquireWriterLock(0);
+
+            try
+            {
+                if (!Expired())
+                {
+                    return;
+                }
+
+                var response = ApiKeyClient.IssueAccessToken(Credentials.ToIssueAccessTokenRequest());
+                AccessToken = response.AccessToken;
+                AccessTokenExpiresAt = response.ExpiresAt;
+                LastException = null;
+            }
+            catch (RpcException e)
+            {
+                if (e.StatusCode == StatusCode.Unauthenticated)
+                {
+                    LastException = e;
+                }
+
+                throw;
+            }
+            catch (Exception e)
+            {
+                throw new Exception($"Failed to issue access token: ${e}");
+            }
+            finally
+            {
+                RWLock.ReleaseWriterLock();
+            }
+        }
+
+        private void ThrowIfUnauthenticated()
+        {
+            RWLock.AcquireReaderLock(0);
+            try
+            {
+                if (LastException is RpcException && LastException.StatusCode == StatusCode.Unauthenticated)
```

Comment:
You're only capturing `LastException` on line 92 if there's an `Unauthenticated` response. So, you could just set a boolean called `unauthenticated` and simplify this line to check that boolean value instead.